kotlinx-coroutines-core/jvm/test/RunBlockingJvmTest.kt

```diff
@@ -12,5 +16,177 @@ class RunBlockingJvmTest : TestBase() {
+            val v = runBlockingNonInterruptible {
+                mayInterrupt()
+                repeat(10) {
+                    expect(it + 1)
+                    delay(1)
+                }
+                42
+            }
+            assertTrue(Thread.interrupted())
+            assertEquals(42, v)
+            expect(11)
+        }
+        finish(12)
+    }
+
+    /**
+     * Tests that [runBlockingNonInterruptible] is going to run its job to completion even if it gets interrupted
+     * or if thread switches occur, and then will rethrow the exception thrown by the job.
+     */
+    @Test
+    fun testNonInterruptibleRunBlockingFailure() {
+        val exception = AssertionError()
+        startInSeparateThreadAndInterrupt { mayInterrupt ->
+            val exception2 = assertFailsWith<AssertionError> {
+                runBlockingNonInterruptible {
+                    mayInterrupt()
+                    repeat(10) {
+                        expect(it + 1)
+                        // even thread switches should not be a problem
+                        withContext(Dispatchers.IO) {
+                            delay(1)
+                        }
+                    }
+                    throw exception
+                }
+            }
+            assertTrue(Thread.interrupted())
+            assertSame(exception, exception2)
+            expect(11)
+        }
+        finish(12)
+    }
+
+
+    /**
+     * Tests that [runBlockingNonInterruptible] is going to run its job to completion even if it gets interrupted
+     * or if thread switches occur.
+     */
+    @Test
+    fun testNonInterruptibleRunBlockingPropagatingInterruptions() {
+        val exception = AssertionError()
+        startInSeparateThreadAndInterrupt { mayInterrupt ->
+            runBlockingNonInterruptible {
+                mayInterrupt()
+                try {
+                    Thread.sleep(Long.MAX_VALUE)
+                } catch (_: InterruptedException) {
+                    expect(1)
+                }
+            }
+            expect(2)
+            assertFalse(Thread.interrupted())
+        }
+        finish(3)
+    }
+
+    /**
+     * Tests that starting [runBlockingNonInterruptible] in an interrupted thread does not affect the result.
+     */
+    @Test
+    fun testNonInterruptibleRunBlockingStartingInterrupted() {
+        Thread.currentThread().interrupt()
+        val v = runBlockingNonInterruptible { 42 }
+        assertEquals(42, v)
+        assertTrue(Thread.interrupted())
+    }
+
+    private fun startInSeparateThreadAndInterrupt(action: (mayInterrupt: () -> Unit) -> Unit) {
+        val latch = CountDownLatch(1)
+        val thread = thread {
+            action { latch.countDown() }
+        }
+        latch.await()
+        thread.interrupt()
+        thread.join()
+    }
+
+    private fun <T> runBlockingNonInterruptible(action: suspend () -> T): T {
+        val result = AtomicReference<Result<T>>()
```

Comment:
Is an atomic ref needed? Seems like it could just be:
```
var result: Result<T> = Result.failure(IllegalStateException("Action did not complete"))
```
Which is a better error than a null pointer if the reference is not set.

Reply:
In this implementation, it's indeed not needed. It would be needed if `runBlockingNonInterruptible` were to accept a `CoroutineContext` (just like the real thing would), allowing the `action` to execute on a separate thread.